docs: add `x/subspaces` spec (#910)
desmos-labs/desmos: x/subspaces/spec/04-messages.md

---
id: messages
title: Messages
sidebar_label: Messages
slug: messages
---

# Msg Service

## Msg/CreateSubspace

> A subspace can be created with the `MsgCreateSubspace` specifying the fields showed below:

Reviewer: ```suggestion
A subspace can be created with the `MsgCreateSubspace`.
```

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L65-L75 
```

The message is expected to fail if one of the following checks inside the function is matched:
```js reference
https://github.com/desmos-labs/desmos/blob/48b7d58e9b09f26639f7a767fbd921dba309350f/x/subspaces/types/models.go#L51-L82

> ```

Reviewer: It's expected to fail if: 
- the provided name is either empty or blank; 
- the specified treasury address (if any) is invalid; 
- the specified owner address (if any) is invalid.

## Msg/EditSubspace
A subspace can be edited with the `MsgEditSubspace`:

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L85-L100 
```

> The message is expected to fail if:
> * The subspace id doesn't exist;
> * The updated subspaces is invalid;
> * The signer has no permission to edit subspace inside the specified subspace.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the updated subspace is invalid;
* the signer has no permission to edit the subspace.
```

## Msg/DeleteSubspace
A subspace can be deleted using `MsgDeleteSubspace`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L105-115 
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The signer has no permission to delete subspace inside the specified subspace.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the signer has no permission to delete the subspace.
```

## Msg/CreateSection
A section can be created using the `MsgCreateSection`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L122-L141 
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The parent section does not exist;
> * The creator has no permission to manage sections inside the specified subspace.

Reviewer: ```suggestion
The message is expected to fail if:
* the subspace does not exist;
* the parent section (if specified) does not exist;
* the creator has no permission to manage sections within the subspace;
* the provided section name is either empty or blank.
```

## Msg/EditSection
A section can be edited using the `MsgEditSection`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L156-L179 
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The section does not exist;
> * The editor has no permission to manage sections inside the specified subspace.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the section does not exist;
* the editor has no permission to manage sections within the subspace.
```

## Msg/MoveSection
A section can be moved to under another section using the `MsgMoveSection`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L184-L207 
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The section does not exist;
> * The destination section does not exist;
> * The updated section is not valid;
> * The signer has no permission to manage sections inside the specified subspace;
> * The section path is invalid.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the section does not exist;
* the destination section does not exist;
* the signer has no permission to manage sections within the subspace;
* the new section path is invalid (this means that is not possible to reach the moved section starting from the root section, or that a circular path is detected).
```

## Msg/DeleteSection
A section can be deleted using the `MsgDeleteSection`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L212-L224 
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The section does not exist;
> * The signer has no permission to manage sections inside the specified subspace.

Reviewer: ```suggestion
It's is expected to fail if:
* the subspace does not exist;
* the section does not exist;
* the signer has no permission to manage sections within the subspace.
```

## Msg/CreateUserGroup
A user group can be created using the `MsgCreateUserGroup`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L235-L261 
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The destination section does not exist;
> * The signer has no permissions to create a user group or set permissions inside the specified section;
> * The permissions values are not valid;
> * The group is not valid;

Reviewer: ```suggestion
The message is expected to fail if:
* the subspace does not exist;
* the section does not exist;
* the signer has no permissions to create a user group or set permissions within the section;
* the permissions values are not valid;
* the provided user group name is either blank or empty.
```

## Msg/EditUserGroup
A user group can be edited using the `MsgEditUserGroup`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L271-L293
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The user group does not exist;
> * The signer has no permission to manage user groups;
> * The updated group is invalid.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the user group does not exist;
* the signer has no permission to manage user groups within the subspace;
* the updated group is invalid.
```

## Msg/MoveUserGroup
A user group can be moved to another section group using the `MsgMoveUserGroup`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L298-L317
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The destination section does not exist;
> * The user group does not exist;
> * The signer has no permission to manage user groups inside the section;
> * The signer has no permissions to manage user groups or set permissions inside the destination section.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the user group does not exist;
* the destination section does not exist;
* the signer has no permission to manage user groups inside the current group's section;
* the signer has no permissions to manage user groups or set permissions inside the destination section.
```

## Msg/SetUserGroupPermissions
A user group permissions can be set using the `MsgSetUserGroupPermissions`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L326-L347
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The user group does not exist;
> * The signer has no permission to set permissions;
> * The permissions values are not valid;
> * The signer is inside the user group and it is not the subspace owner.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the user group does not exist;
* the signer has no permission to set permissions within the group's section;
* the permissions values are not valid;
* the signer is inside the user group and it is not the subspace owner.
```

## Msg/DeleteUserGroup
A user group permissions can be deleted using the `MsgDeleteUserGroup`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L353-L369
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The user group does not exist;
> * The signer has no permission to manage sections inside the specified section.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the user group does not exist;
* the signer has no permission to manage sections inside the group's section.
```

## Msg/AddUserToUserGroup
A user can be added to a user group using the `MsgAddUserToUserGroup`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L376-L396
```

It's expected to fail if:

> * The subspace does not exist;
> * The user group does not exist;
> * The signer has no permission to set permissions inside the subspace and section where user group live;
> * The sender already is the member of the user group.

Reviewer: ```suggestion
* the subspace does not exist;
* the user group does not exist;
* the signer has no permission to set permissions inside the subspace and section where user group is;
* the user already is a member of the user group.
```

## Msg/RemoveUserFromUserGroup
A user can be removed from a user group using the `MsgRemoveUserFromUserGroup`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L402-L422
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The user group does not exist;
> * The sender has no permission to set permissions inside the subspace and section where user group live;
> * The user is not the member of the user group.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the user group does not exist;
* the sender has no permission to set permissions inside the subspace and section where user group is;
* the user is not the member of the user group.
```

## Msg/SetUserPermissions
A user permissions can be set using the `MsgSetUserPermissions`.

```js reference
https://github.com/desmos-labs/desmos/blob/master/proto/desmos/subspaces/v2/msgs.proto#L430-L454
```

> The message is expected to fail if:
> * The subspace does not exist;
> * The section does not exist;
> * The signer has no permission to set permissions inside the destination section;
> * The permissions values are not valid.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the section does not exist;
* the signer has no permission to set permissions inside the destination section;
* the permissions values are not valid.
```
